lib/elixir/lib/calendar/iso.ex

```diff
@@ -1455,6 +1455,134 @@ defmodule Calendar.ISO do
     {days, {@parts_per_day - 1, @parts_per_day}}
   end
 
+  @doc """
+  Shifts date by Calendar.Duration units according to its calendar.
+
+  Available units are: `:year, :month, :week, :day, :hour, :minute, :second, :microsecond`.
+
+  When shifting by month:
+  - it will shift to the current day of a month
+  - when the current day does not exist in a month, it will shift to the last day of a month
+
+  ## Examples
+
+      iex> Calendar.ISO.shift_date(2016, 1, 3, month: 2)
+      {2016, 3, 3}
+      iex> Calendar.ISO.shift_date(2016, 2, 29, month: 1)
+      {2016, 3, 29}
+      iex> Calendar.ISO.shift_date(2016, 1, 31, month: 1)
+      {2016, 2, 29}
+      iex> Calendar.ISO.shift_date(2016, 1, 31, year: 4, day: 1)
+      {2020, 2, 1}
+  """
+  @spec shift_date(year(), month(), day(), [Calendar.Duration.duration_unit()]) ::
+          {year, month, day}
+  @impl true
+  def shift_date(year, month, day, duration_units) do
+    {year, month, day, _, _, _, _} =
+      shift_naive_datetime(year, month, day, 0, 0, 0, {0, 0}, duration_units)
```

Comment:
I would implement this separately, because if any of the time duration are non-zero, we should raise.

Reply:
do we really want `Date.shift(~D[2024-01-01], second: 864001)` to raise? if so, wouldn't it still be beneficial to use `shift_naive_datetime/8` like so:

```
case shift_naive_datetime(year, month, day, 0, 0, 0, {0, 0}, duration_units) do
  {year, month, day, 0, 0, 0, {0, 0}} -> {year, month, day}
 _ -> raise "..."
end
```

Reply:
I would have two distinct loops and then we reuse the internal functions (such a shift_numerical, shift_month, etc). Especially because i don't think we should convert days to seconds in the date case.